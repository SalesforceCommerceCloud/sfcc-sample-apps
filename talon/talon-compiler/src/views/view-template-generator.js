const { moduleSpecifierToElementName, moduleSpecifierToId, convertToKebabCase } = require('talon-common/dist/cjs');
const { generateAttributesJS, DEFAULT_ATTRIBUTES } = require('./attributes-js-generator');

const pretty = require("pretty");
const ATTRIBUTE_REFERENCE = "attributes";

function regions(regionList, isThemeLayout, attributeSet, isRenderDesignTime) {
    let html = '';
    if (regionList && regionList.length) {
        regionList.forEach(region => {
            let regionHtml = '';
            if (region.components && region.components.length) {
                region.components.forEach(cmp => {
                    regionHtml += isRenderDesignTime ? getDropZone() : '';
                    regionHtml += component(cmp, false, region.name, attributeSet, false, isRenderDesignTime).html;
                });
                regionHtml += isRenderDesignTime ? getDropZone() : '';
            }

            html += isRenderDesignTime ? getRegionWrapper(region.name, regionHtml) : regionHtml;
        });
    }

    return `${html}
    ${isThemeLayout ? defaultSlot() : ""}`;
}

function getRegionWrapper(regionName, regionHtml) {
    const slotAttribute = regionName ? `slot="${regionName}"` : ``;
    return `<talondesign-region-wrapper ${slotAttribute}>${regionHtml}</talondesign-region-wrapper>`;
}

function getDropZone() {
    return `<talondesign-drop-region></talondesign-drop-region>`;
}

function defaultSlot() {
    return `<talon-router-container></talon-router-container>`;
}

function attributes(cmp, attributeSet, cmpNameKey) {
    if (!cmp.attributes) {
        return {html: '', attributes: attributeSet};
    }

    let attrHtml = '';
    Object.entries(cmp.attributes).forEach(([attributeName, attributeValue]) => {
        attrHtml += ` ${convertToKebabCase(attributeName)}={${ATTRIBUTE_REFERENCE}.${cmpNameKey}.${attributeName}}`;
        // Store attribute value for runtime injection
        attributeSet[cmpNameKey] = attributeSet[cmpNameKey] || {};
        attributeSet[cmpNameKey][attributeName] = attributeValue;
    });

    return {html: attrHtml, attributes: attributeSet};
}

function buildSlot(slotName, isRenderDesignTime) {
    return `${slotName && !isRenderDesignTime ? ` slot="${slotName}"` : ""} `;
}

function component(cmp, isThemeLayout, regionName, attributeSet = {}, isTopLevel, isRenderDesignTime) {
    const isRenderWrapper = !isTopLevel && isRenderDesignTime;
    const elem = moduleSpecifierToElementName(cmp.name);
    const cmpNameKey = getCmpNameKey(cmp, attributeSet);
    const attrs = attributes(cmp, attributeSet, cmpNameKey);
    const slot = buildSlot(regionName, isRenderDesignTime);
    const rgns = regions(cmp.regions, isThemeLayout, attrs.attributes, isRenderDesignTime);

    const cmpHtml = `<${elem}${attrs.html}${slot}>${rgns}</${elem}>`;

    const html = isRenderWrapper ? getComponentWrapper(cmp, cmpNameKey) : cmpHtml;

    return { html, attributes: attrs };
}

function getCmpNameKey(cmp, attributeSet) {
    const componentName = moduleSpecifierToId(cmp.name);
    const globalCmpIdx = Object.keys(attributeSet).length;
    return [componentName, globalCmpIdx].join("_");
}

function getComponentWrapper(cmp, cmpNameKey) {
    const attributesAttr = cmp.attributes ? `attributes={${ATTRIBUTE_REFERENCE}.${cmpNameKey}}` : '';
    return `<talondesign-component-wrapper ${attributesAttr} name="${cmp.name}"></talondesign-component-wrapper>`;
}

/**
 * Generate an LWC template for a provided component
 *
 * @param {object*} cmp - The top level component to render an LWC template for
 * @param {boolean} isThemeLayout - Whether that component is a theme layout or not
 */
function template(cmp, isThemeLayout, isRenderDesignTime = false) {
    const result = component(cmp, isThemeLayout, null, {}, true, isRenderDesignTime);
    const attributesVal = generateAttributesJS(result.attributes.attributes);
    return { "html": pretty(`<template>${result.html}</template>`, {ocd: true}), "attributes": attributesVal };
}

/**
 * Generates javascript which imports an HTML file and exports the HTML as default
 *
 * @param {string} moduleName - The name of the view, used to specify the import location of the HTML
 * @param {string} attributesVal - The attribute set value of the view, used to inject component attribute values during runtime
 */
function javascript(moduleName, attributesVal) {
    return `import html from './${moduleName}.html'
    export default {
        "html": html,
        "attributes": ${attributesVal || DEFAULT_ATTRIBUTES}
    }`;
}

module.exports = { template, javascript };