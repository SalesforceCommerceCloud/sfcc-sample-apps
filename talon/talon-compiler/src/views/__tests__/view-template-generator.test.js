import { template, javascript } from '../view-template-generator.js';

const defaultView = {
    name: "defaultView",
    component: {
        name: "x/foo",
        regions: [
            {
                name: "header",
                components: [
                    {
                        name: "x/header",
                        attributes: {
                            label: "Header Label",
                            class: "header"
                        }
                    },
                    {
                        name: "x/leftSide",
                        regions: [
                            {
                                name: "left-side",
                                components: [
                                    {
                                        name: "x/logo",
                                        attributes: {}
                                    }
                                ]
                            }
                        ]
                    }
                ]
            },
            {
                name: "customHeader"
            },
            {
                name: "footer",
                components: [
                    {
                        name: "x/footer",
                        attributes: {
                            label: "Label",
                            number: -3.14,
                            recordId: "{!recordId}",
                        }
                    }
                ]
            }
        ]
    }
};

describe("view-template-generator", () => {
    describe("template", () => {
        it("generates the right view html", () => {
            const generatedComponent = template(defaultView.component, false);
            expect(generatedComponent.html).toMatchSnapshot();
        });
        it("generates the right view attributes", () => {
            const generatedComponent = template(defaultView.component, false);
            expect(generatedComponent.attributes).toMatchSnapshot();
        });
        it("generates the right theme layout view html", () => {
            const generatedComponent = template(defaultView.component, true);
            expect(generatedComponent.html).toMatchSnapshot();
        });
        it("generates the right theme layout view attributes", () => {
            const generatedComponent = template(defaultView.component, true);
            expect(generatedComponent.attributes).toMatchSnapshot();
        });

        it("generates the right view html for designtime", () => {
            expect(template(defaultView.component, false, true)).toMatchSnapshot();
        });

        it("generates the right theme layout view html for designtime", () => {
            expect(template(defaultView.component, true, true)).toMatchSnapshot();
        });
    });

    describe("javascript", () => {
        it("generates the right javascript", () => {
            const mockAttributesVal = "MOCK_GENERATED_ATTRIBUTES_JS";
            const generatedJavascript = javascript(defaultView.name, mockAttributesVal);
            expect(generatedJavascript).toMatchSnapshot();
        });
        it("generates the right javascript with no attributes", () => {
            const generatedJavascript = javascript(defaultView.name);
            expect(generatedJavascript).toMatchSnapshot();
        });
    });
});