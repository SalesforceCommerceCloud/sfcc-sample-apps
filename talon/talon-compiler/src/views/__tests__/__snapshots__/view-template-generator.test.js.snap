// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view-template-generator javascript generates the right javascript 1`] = `
"import html from './defaultView.html'
    export default {
        \\"html\\": html,
        \\"attributes\\": MOCK_GENERATED_ATTRIBUTES_JS
    }"
`;

exports[`view-template-generator javascript generates the right javascript with no attributes 1`] = `
"import html from './defaultView.html'
    export default {
        \\"html\\": html,
        \\"attributes\\": (cmp) => ({})
    }"
`;

exports[`view-template-generator template generates the right theme layout view attributes 1`] = `
"(cmp) => ({
    \\"xheader_0\\": {
        \\"label\\": \\"Header Label\\",
        \\"class\\": \\"header\\"
    },
    \\"xfooter_1\\": {
        \\"label\\": \\"Label\\",
        \\"number\\": -3.14,
        \\"recordId\\": (cmp.routeParams[\\"recordId\\"] || '')
    }
})"
`;

exports[`view-template-generator template generates the right theme layout view html 1`] = `
"<template>
  <x-foo>
    <x-header label={attributes.xheader_0.label} class={attributes.xheader_0.class} slot=\\"header\\">
    </x-header>
    <x-left-side slot=\\"header\\">
      <x-logo slot=\\"left-side\\">
      </x-logo>
    </x-left-side>
    <x-footer label={attributes.xfooter_1.label} number={attributes.xfooter_1.number} record-id={attributes.xfooter_1.recordId} slot=\\"footer\\">
    </x-footer>
    <talon-router-container></talon-router-container>
  </x-foo>
</template>"
`;

exports[`view-template-generator template generates the right theme layout view html for designtime 1`] = `
Object {
  "attributes": "(cmp) => ({
    \\"xheader_0\\": {
        \\"label\\": \\"Header Label\\",
        \\"class\\": \\"header\\"
    },
    \\"xfooter_1\\": {
        \\"label\\": \\"Label\\",
        \\"number\\": -3.14,
        \\"recordId\\": (cmp.routeParams[\\"recordId\\"] || '')
    }
})",
  "html": "<template>
  <x-foo>
    <talondesign-region-wrapper slot=\\"header\\">
      <talondesign-drop-region></talondesign-drop-region>
      <talondesign-component-wrapper attributes={attributes.xheader_0} name=\\"x/header\\"></talondesign-component-wrapper>
      <talondesign-drop-region></talondesign-drop-region>
      <talondesign-component-wrapper name=\\"x/leftSide\\"></talondesign-component-wrapper>
      <talondesign-drop-region></talondesign-drop-region>
    </talondesign-region-wrapper>
    <talondesign-region-wrapper slot=\\"customHeader\\"></talondesign-region-wrapper>
    <talondesign-region-wrapper slot=\\"footer\\">
      <talondesign-drop-region></talondesign-drop-region>
      <talondesign-component-wrapper attributes={attributes.xfooter_1} name=\\"x/footer\\"></talondesign-component-wrapper>
      <talondesign-drop-region></talondesign-drop-region>
    </talondesign-region-wrapper>
    <talon-router-container></talon-router-container>
  </x-foo>
</template>",
}
`;

exports[`view-template-generator template generates the right view attributes 1`] = `
"(cmp) => ({
    \\"xheader_0\\": {
        \\"label\\": \\"Header Label\\",
        \\"class\\": \\"header\\"
    },
    \\"xfooter_1\\": {
        \\"label\\": \\"Label\\",
        \\"number\\": -3.14,
        \\"recordId\\": (cmp.routeParams[\\"recordId\\"] || '')
    }
})"
`;

exports[`view-template-generator template generates the right view html 1`] = `
"<template>
  <x-foo>
    <x-header label={attributes.xheader_0.label} class={attributes.xheader_0.class} slot=\\"header\\">
    </x-header>
    <x-left-side slot=\\"header\\">
      <x-logo slot=\\"left-side\\">
      </x-logo>
    </x-left-side>
    <x-footer label={attributes.xfooter_1.label} number={attributes.xfooter_1.number} record-id={attributes.xfooter_1.recordId} slot=\\"footer\\">
    </x-footer>
  </x-foo>
</template>"
`;

exports[`view-template-generator template generates the right view html for designtime 1`] = `
Object {
  "attributes": "(cmp) => ({
    \\"xheader_0\\": {
        \\"label\\": \\"Header Label\\",
        \\"class\\": \\"header\\"
    },
    \\"xfooter_1\\": {
        \\"label\\": \\"Label\\",
        \\"number\\": -3.14,
        \\"recordId\\": (cmp.routeParams[\\"recordId\\"] || '')
    }
})",
  "html": "<template>
  <x-foo>
    <talondesign-region-wrapper slot=\\"header\\">
      <talondesign-drop-region></talondesign-drop-region>
      <talondesign-component-wrapper attributes={attributes.xheader_0} name=\\"x/header\\"></talondesign-component-wrapper>
      <talondesign-drop-region></talondesign-drop-region>
      <talondesign-component-wrapper name=\\"x/leftSide\\"></talondesign-component-wrapper>
      <talondesign-drop-region></talondesign-drop-region>
    </talondesign-region-wrapper>
    <talondesign-region-wrapper slot=\\"customHeader\\"></talondesign-region-wrapper>
    <talondesign-region-wrapper slot=\\"footer\\">
      <talondesign-drop-region></talondesign-drop-region>
      <talondesign-component-wrapper attributes={attributes.xfooter_1} name=\\"x/footer\\"></talondesign-component-wrapper>
      <talondesign-drop-region></talondesign-drop-region>
    </talondesign-region-wrapper>
  </x-foo>
</template>",
}
`;
